firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/Mutation.java

```diff
@@ -191,17 +191,51 @@ static SnapshotVersion getPostMutationVersion(MutableDocument document) {
    * @return A map of fields to transform results.
    */
   protected Map<FieldPath, Value> localTransformResults(
-      Timestamp localWriteTime, MutableDocument mutableDocument) {
+      Timestamp localWriteTime, MutableDocument mutableDocument, Mutation mutation) {
     Map<FieldPath, Value> transformResults = new HashMap<>(fieldTransforms.size());
     for (FieldTransform fieldTransform : fieldTransforms) {
       TransformOperation transform = fieldTransform.getOperation();
       Value previousValue = mutableDocument.getField(fieldTransform.getFieldPath());
+      if (mutation != null) {
+        FieldUpdate update = getFieldUpdate(fieldTransform.getFieldPath());
+        if (update.type == FieldUpdate.Type.DELETE) {
+          previousValue = null;
+        } else if (update.type == FieldUpdate.Type.SET) {
+          previousValue = update.value;
+        }
+      }
       transformResults.put(
           fieldTransform.getFieldPath(), transform.applyToLocalView(previousValue, localWriteTime));
     }
     return transformResults;
   }
 
+  protected static class FieldUpdate {
```

Comment:
This is pretty elegant but maybe we don't need this if we use an ObjectValue rather than MutableDocument+Mutation strategy as described above.

Reply:
See above.